### ISSUE TITLE
[FIX-017] Preserve NNModel subclasses through INT8 quantization

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#227]
**Category:** Numerics and quantization · **Priority:** P2 · **Size:** S

**Evidence status:** Reproduced subtype-composition defect.

## Outcome and scope

INT8 quantization must return a sibling of the original NNModel subtype so generation methods and custom overrides survive. Preserve the existing shallow-sharing contract for non-network attributes and the independent deep-copied quantized net. Do not rerun constructors, clone tokenizers or change quantization/export algorithms.

## Current behavior / evidence

**Context and reproduction:** Build a standard `GenerativeNNModel` with a small TransformerNN and call `quantize_int8(model)`. The original type is `GenerativeNNModel`, the returned type is `NNModel`, and `hasattr(result, 'generate')` is false. Tokenizer data is copied into the result's dict but its generation behavior is lost. Other subclasses likewise lose methods/overrides despite having ordinary nn.Linear nets accepted by the function.

## Repository map

- [`src/nnx/quantize/ptq.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/quantize/ptq.py#L87) — Hardcoded base-type allocation discards subtype behavior after quantization.
- [`src/nnx/nn/generative_nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/generative_nn_model.py#L34) — Tokenizer-bearing subclass and `generate` contract.
- [`tests/test_quantize_ptq.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_quantize_ptq.py#L71) — Current base-type check is insufficient; module gates on torchao.
- [`tests/test_quantize_ptq.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_quantize_ptq.py#L119) — Established shared params/device/loss ownership.
- [`tests/test_generative_nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_generative_nn_model.py#L35) — Tiny local BPE/model fixtures require no model download.
- [`pyproject.toml`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/pyproject.toml#L121) — Optional quantize and lm extras supply runtime test dependencies.

## Implementation approach

1. Retain `deepcopy(model.net)` and the existing `quantize_(..., Int8WeightOnlyConfig(version=2))` call. Replace the hardcoded NNModel allocation/dict copy with an appropriate `copy.copy(model)` sibling, then assign only its `.net` to the independent quantized copy.
2. Assert exact subtype identity, distinct wrapper identity and distinct network identity. Params/net_params/loss_fn/tokenizer and ordinary custom attributes continue to follow the existing shallow-sharing contract. Do not call `type(model)(...)` or `__init__`, which can require extra arguments, initialize new weights and advance RNG.
3. Add a small ordinary NNModel subclass with an overridden method and an extra attribute, plus a real GenerativeNNModel with the existing tiny local tokenizer recipe. This bounds support to normal shallow-copy-compatible subclasses; arbitrary custom copy hooks need not be silently bypassed.
4. Keep optional-import errors, INT8 numerical/size coverage and existing ONNX paths unchanged. Quantization still does not imply supported training of the quantized wrapper. The original network’s parameter values and parameter classes must remain unchanged.

## Acceptance criteria

- [ ] Quantizing a GenerativeNNModel returns a GenerativeNNModel with its tokenizer and callable `generate` intact.
- [ ] A deterministic small-tokenizer generation smoke test completes after quantization and respects the token limit.
- [ ] A custom NNModel subclass retains an overridden method and custom instance attribute.
- [ ] Existing base-class output/size/ONNX tests pass and the source network remains unquantized and unchanged.

## Validation and completion

Self-contained subtype fixture; requires the existing `quantize` extra:

```python
from nnx import Devices, Losses, Nets, NNModel, NNModelParams, NNParams, quantize_int8

class TaggedModel(NNModel):
    def label(self):
        return self.tag

model = TaggedModel(
    net_params=NNParams(input_dim=4, output_dim=2, hidden_dims=[8], dropout_prob=0),
    params=NNModelParams(net=Nets.FEED_FWD, device=Devices.CPU,
                         loss=Losses.CROSS_ENTROPY),
)
model.tag = "original-subtype"
quantized = quantize_int8(model)
assert type(quantized) is TaggedModel
assert quantized.label() == "original-subtype"
assert quantized is not model and quantized.net is not model.net
assert quantized.params is model.params
```

Add `test_quantize_int8_preserves_custom_subclass` and `test_quantized_generative_model_can_generate`. For generation, reuse the local five-sentence BPE setup in `tests/test_generative_nn_model.py`, build the tiny Transformer, quantize, and call `generate(prompt="the", max_new_tokens=3, temperature=0.0)`. Assert tokenizer identity, callable generate, deterministic repeated output and at most three newly emitted token callbacks; count emitted IDs directly rather than relying only on re-tokenized decoded text. Also run cache/no-cache supported paths, source value/type snapshots and existing base-class size/output/ONNX checks. This ticket requires installed torchao and tokenizers in the actual test lane; an entirely skipped quantization module is not completion evidence.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_quantize_ptq.py tests/test_generative_nn_model.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Public subtype contract.** Fixing [quantize_int8](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/quantize/ptq.py) also requires its public type signature and docstring to preserve a caller’s NNModel subtype. Use a Python-3.10-compatible bounded TypeVar or equivalent supported typing so `quantize_int8(generative).generate(...)` type-checks without a cast. Copy wrapper state using the chosen shallow-copy semantics without rerunning constructors, while deep-copying only `net`; preserve tokenizer/config/custom attributes and ordinary subclass copy behavior. Exercise both [quantize facade](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/quantize/__init__.py) and top-level `nnx` exports, with no eager optional torchao import.

- **Generation and persistence boundaries.** [generate](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/generative_nn_model.py) consumes tokenizer, `net_params.max_seq_len`, cache helpers, token callbacks and the full mode map. Test short greedy cached and full-recompute generation on the quantized result, callback counts/token limit, and mode restoration, then call the original FP32 model again to prove isolation. Quantized and original text need not be identical because weight rounding can change argmax. Retain existing base-class PTQ prediction/size/ONNX coverage, but do not infer new raw-state or Hub reconstruction recipes from preserving the in-memory wrapper class; those are independent artifact contracts.

- **Bounded optional-extra example.** [Example 12](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/12_quantize_int8.py) currently teaches classifier quantization and a full ONNX export workload. Add a separate tiny generative helper using a locally trained tokenizer and a one-layer Transformer, with at most two new tokens on each decode path. Guard/lazily import tokenizer and quantization extras inside execution so normal example import remains offline and valid without them. Execute this helper with `[quantize,lm]` in the current supported torchao environment; it need not install ONNX merely to test subtype preservation. Update the return-value description and point LM users to this concrete inference-only composition.

### Samples and executable coverage

- [`examples/12_quantize_int8.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/12_quantize_int8.py): Add a lazy-extra tiny GenerativeNNModel helper with local BPE training, quantization, cached/no-cache two-token generation, callback/mode/source-isolation assertions; keep classifier/ONNX main separate. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k quantized_generative_subtype. Current supported torchao plus tokenizers extras; 1 layer, d_model=16, short local prompt and at most 2 generated tokens per path, no training/download/ONNX requirement. Missing extras are explicit skips, not passes. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`README.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/README.md) | `generated/site/Repository-Overview.md` | `generated/wiki/Repository-Overview.md` | State that quantize_int8 returns a fresh wrapper of the input NNModel subtype with an independent quantized net and shared non-net attributes, rather than always a base NNModel. |
| [`docs/lm.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/lm.md) | `generated/site/Language-modeling.md` | `generated/wiki/Language-modeling.md` | Add an inference-only quantize_int8 → generate example with optional quantize/lm extras, cache/no-cache and token-limit behavior, source-model isolation and no new generic persistence promise. |
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Clarify PTQ wrapper shallow-copy/net deep-copy ownership and preserved subtype methods/attributes; retain existing base-model export support boundaries. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Public top-level and quantize-facade calls preserve the concrete wrapper subtype in runtime behavior and supported static typing, allowing generate on the returned generative model without a cast or constructor rerun.
- [ ] The quantized generative wrapper completes cached and full-recompute bounded decoding with correct token callback counts and restored child modes; calling the original model afterwards proves its network and generation behavior were not mutated.
- [ ] Subtype retention does not alter existing PTQ checkpoint/export assertions or claim a new Hub/raw-state reconstruction recipe; tokenizer and other non-net attributes follow the documented shallow-copy ownership.
- [ ] The bounded Example 12 generative helper executes with the declared optional extras using only local tokenizer data, and verifies subtype, decoding limits, mode restoration and source isolation separately from classifier import smoke.
- [ ] The issue-specific changes in `README.md`, `docs/lm.md`, `docs/concepts.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FEAT-016](https://github.com/thekaveh/NNx/issues/238).

## Delivery order and dependencies

- **Execution order:** 28 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.

**Dependency rationale.** Independent repair of current behavior. No feature dependency; use installed optional torchao in relevant test lane.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `quantize_int8` allocated a bare `NNModel.__new__(NNModel)` and copied the source `__dict__` into it, so a `GenerativeNNModel` came back as a plain `NNModel` (tokenizer data copied, `generate` gone) and user subclasses lost their overrides.

**What changed**
- The wrapper is now `copy.copy(model)` with `.net` replaced by the quantized deep copy: exact subtype kept, distinct wrapper and net identities, shallow-shared `params` / `net_params` / `loss_fn` / `tokenizer` / custom attributes, no constructor re-run (no re-initialized weights, no RNG consumed). Signature typed `quantize_int8(model: _M) -> _M` with a bounded `TypeVar` so `quantize_int8(generative).generate(...)` type-checks. Quantization call, optional-import error, INT8 numerics and ONNX paths unchanged.
- Example 12: bounded `quantized_generative_subtype()` (`tests/test_examples_smoke.py -k quantized_generative_subtype`) — lazy `[quantize,lm]` imports raising `ImportError` naming the extra (the smoke harness now reports that as an explicit skip, never a pass); one-layer `d_model=16` `GenerativeNNModel` with a locally trained BPE tokenizer, ≤2 new tokens on cached and full-recompute paths, callback bound, mode restoration, FP32-source isolation. Classifier/ONNX `main` untouched.
- Docs: README + PyPI readme, `concepts.md` quantize bullet, examples catalog, CHANGELOG, API reference.

**PRs:** #279 (develop, merged), #280 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1626 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_quantize_int8_preserves_custom_subclass`, `test_quantized_generative_model_can_generate` (torchao 0.17.0 + tokenizers 0.22.2 lane — executed, not skipped); existing base-class PTQ output/size/attr/ONNX tests unchanged.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / `build_pypi_readme --check` / projection tests) → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #279 and #280.

**Not verifiable locally:** none; live Pages/wiki publication is left to the `docs.yml` main-push job.
